rmw_fastrtps_shared_cpp/src/rmw_subscription.cpp

```diff
@@ -39,25 +39,12 @@ __rmw_destroy_subscription(
   const rmw_node_t * node,
   rmw_subscription_t * subscription)
 {
-  if (!node) {
-    RMW_SET_ERROR_MSG("node handle is null");
-    return RMW_RET_ERROR;
-  }
-
-  if (node->implementation_identifier != identifier) {
-    RMW_SET_ERROR_MSG("node handle not from this implementation");
-    return RMW_RET_ERROR;
-  }
-
-  if (!subscription) {
-    RMW_SET_ERROR_MSG("subscription handle is null");
-    return RMW_RET_ERROR;
-  }
-  if (subscription->implementation_identifier != identifier) {
-    RMW_SET_ERROR_MSG("subscription handle not from this implementation");
-    return RMW_RET_ERROR;
-  }
+  assert(node->implementation_identifier == identifier);
+  assert(subscription->implementation_identifier == identifier);
```

Comment:
An `assert` is not the same as the checks it is replacing... Why is this being changed?

Reply:
This was discussed elsewhere (https://github.com/ros2/rmw_fastrtps/pull/414#discussion_r460270498 comes to mind). Long story short, it doesn't make much sense to perform the same checks twice. So here those (redundant) checks in `rmw_fastrtps_shared_cpp` are replaced by assertions, and both `rmw_fastrtps_cpp` and `rmw_fastrtps_dynamic_cpp`  are expected to make proper use of this public but internal API.

Reply:
Makes sense, but wouldn't it make more sense to check here once? Rather than checking in each and only asserting here? I'm guessing that wasn't possible.

Reply:
We do that in many cases, but sometimes `rmw_fastrtps_cpp` or `rmw_fastrtps_dynamic_cpp` don't just forward the call to `rmw_fastrtps_shared_cpp` and thus validation has to happen there before doing anything vendor specific.